File: simtest/simtest/orchestrator.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
import json

from simtest.boto_session import get_session

_orchestrator = None


def get_orchestrator():
    if not _orchestrator:
        raise Exception("Orchestrator has not been initialized")
    return _orchestrator


def create_orchestrator(region):
    global _orchestrator
    _orchestrator = Orchestrator(region)
    return _orchestrator


class Orchestrator:
    def __init__(self, region):
        self._session = get_session()
        self._region = region
        self._arn = f"arn:{self._session.get_partition()}:states:{self._region}:{self._session.get_account()}:stateMachine:SO0111-SHARR-Orchestrator"

    def invoke(self, payload):
        try:
            sfn = self._session.client("stepfunctions", region_name=self._region)
            print(f"Invoking Orchestrator in {self._region}")
            sfn.start_execution(stateMachineArn=self._arn, input=json.dumps(payload))
        except Exception as e:
            print(e)
            print(
                f"start_execution for Orchestrator step function failed in {self._region}"
            )

    def get_region(self):
        return self._region

File: source/lambdas/common/utils/orchestrator.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { SFNClient, StartExecutionCommand } from '@aws-sdk/client-sfn';
import { Logger } from '@aws-lambda-powertools/logger';

const ORCHESTRATOR_ARN = process.env.ORCHESTRATOR_ARN;

export function createSFNClient(): SFNClient {
  return new SFNClient({ maxAttempts: 10 });
}

/**
 * Executes the orchestrator Step Function with the provided input
 * @param input - The JSON string input for the Step Function
 * @param logger - Logger instance for logging
 * @param sfnClient - Optional SFN client instance (will create one if not provided)
 * @returns The execution ID of the started Step Function execution or undefined if extraction fails
 */
export async function executeOrchestrator(
  input: string,
  logger: Logger,
  sfnClient?: SFNClient,
): Promise<string | undefined> {
  if (!ORCHESTRATOR_ARN) {
    throw new Error('ORCHESTRATOR_ARN env is required');
  }

  const client = sfnClient || createSFNClient();

  const response = await client.send(
    new StartExecutionCommand({
      stateMachineArn: ORCHESTRATOR_ARN,
      input,
    }),
  );

  const executionArn = response.executionArn;

  if (!executionArn) {
    logger.warn('Failed to get execution ARN from response', {
      executionArn: response.executionArn,
    });
    return undefined;
  }

  logger.debug(`Orchestrator triggered on input ${input}`, { executionArn });

  return executionArn;
}
